src/coordination/coordinator_instance.cpp

```diff
@@ -27,66 +27,97 @@ using nuraft::ptr;
+          [this] { std::ranges::for_each(repl_instances_, &ReplicationInstance::StopFrequentCheck); })) {
+  auto find_repl_instance = [](CoordinatorInstance *self,
+                               std::string_view repl_instance_name) -> ReplicationInstance & {
+    auto repl_instance =
+        std::ranges::find_if(self->repl_instances_, [repl_instance_name](ReplicationInstance const &instance) {
+          return instance.InstanceName() == repl_instance_name;
+        });
+
+    MG_ASSERT(repl_instance != self->repl_instances_.end(), "Instance {} not found during callback!",
+              repl_instance_name);
+    return *repl_instance;
   };
 
-  replica_succ_cb_ = [find_instance](CoordinatorInstance *coord_instance, std::string_view instance_name) -> void {
-    auto lock = std::lock_guard{coord_instance->coord_instance_lock_};
-    spdlog::trace("Instance {} performing replica successful callback", instance_name);
-    find_instance(coord_instance, instance_name).OnSuccessPing();
+  replica_succ_cb_ = [find_repl_instance](CoordinatorInstance *self, std::string_view repl_instance_name) -> void {
+    auto lock = std::lock_guard{self->coord_instance_lock_};
+    spdlog::trace("Instance {} performing replica successful callback", repl_instance_name);
+    auto &repl_instance = find_repl_instance(self, repl_instance_name);
+
+    if (!repl_instance.EnsureReplicaHasCorrectMainUUID(self->GetMainUUID())) {
+      spdlog::error(
+          fmt::format("Failed to swap uuid for replica instance {} which is alive", repl_instance.InstanceName()));
+      return;
+    }
+
+    repl_instance.OnSuccessPing();
   };
 
-  replica_fail_cb_ = [find_instance](CoordinatorInstance *coord_instance, std::string_view instance_name) -> void {
-    auto lock = std::lock_guard{coord_instance->coord_instance_lock_};
-    spdlog::trace("Instance {} performing replica failure callback", instance_name);
-    find_instance(coord_instance, instance_name).OnFailPing();
+  replica_fail_cb_ = [find_repl_instance](CoordinatorInstance *self, std::string_view repl_instance_name) -> void {
+    auto lock = std::lock_guard{self->coord_instance_lock_};
+    spdlog::trace("Instance {} performing replica failure callback", repl_instance_name);
+    auto &repl_instance = find_repl_instance(self, repl_instance_name);
+    repl_instance.OnFailPing();
+    // We need to restart main uuid from instance since it was "down" at least a second
+    // There is slight delay, if we choose to use isAlive, instance can be down and back up in less than
+    // our isAlive time difference, which would lead to instance setting UUID to nullopt and stopping accepting any
+    // incoming RPCs from valid main
+    // TODO(antoniofilipovic) this needs here more complex logic
+    // We need to get id of main replica is listening to on successful ping
+    // and swap it to correct uuid if it failed
+    repl_instance.ResetMainUUID();
```

Comment:
It seems to me that REPLICA's connection has nothing to do with MAIN's UUID.
UUID is changed during Failover, so that should invalidate the REPLICA's state, not this.

Reply:
agree but there is some edge case which fico wanted to cover with this code. This code exists from before I just renamed function to more explicit

Reply:
I work on that here: https://github.com/memgraph/memgraph/pull/1711